examples/small-coalton-programs/src/brainfold.lisp

```diff
@@ -0,0 +1,341 @@
+
+  ;;
+  ;; Accessing Brainfold cells using pointer
+  ;;
+  
+  (define (nth-bf-cell index)
+    "Finds the vector entry associated with the given index."
+    (vec:index index bf-array))
+
+  (define (current-bf-cell-value)
+    "Returns the value of the bf-cell indexed by the pointer."
+    (unwrap (nth-bf-cell (cell:read pointer))))
+
+  ;;
+  ;; Print Buffer for printing
+  ;;
+
+  (define print-buffer (cell:new ""))
+  
+  ;;
+  ;; Clean Environment
+  ;;
+  
+  (define (reset-bf)
+    "Resets the bf-environment- pointer, print-buffer, and array."
+    (cell:write! pointer 0)
+    (cell:write! print-buffer "")
+    (reset-bf-array)))
+
+;;;
+;;; Commands (Functions called by Brainfold Cmds)
+;;;
+
+(coalton-toplevel
+
+  ;;
+  ;; Navigating through bf-cells (> <)
+  ;;
+  
+  (define (move-right)
+    "Moves the pointer one bf-cell to the right."
+    (cell:increment! pointer))
+
+  (define (move-left)
+    "Moves the pointer one bf-cell to the left."
+    (cell:decrement! pointer))
+
+  ;;
+  ;; Changing bf-cell values (+ -)
+  ;;
+
+  (define (incr)
+    "Increments the value for the current bf-cell."
+    (vec:set! (cell:read pointer)
+              (+ 1 (current-bf-cell-value))
+              bf-array)
+    (current-bf-cell-value))
+
+  (define (decr)
+    "Decrements the value for the current bf-cell."
+    (vec:set! (cell:read pointer)
+              (- (current-bf-cell-value) 1)
+              bf-array)
+    (current-bf-cell-value))
+
+  ;;
+  ;; Printing Cells (.)
+  ;;
+  
+  (declare print (Ufix -> String))
+  (define (print ascii-value)
+    "Converts a Ufix into its ascii value."
+    (into (make-list (unwrap (char:code-char ascii-value)))))
+
+  
+  (define (print-current-cell)
+    "Prints the current bf-cell."
+    (cell:write! print-buffer
+                 (str:concat (cell:read print-buffer)
+                             (print
+                              (unwrap
+                               (tryinto
+                                (current-bf-cell-value))))))
+    (current-bf-cell-value))
+
+  ;;
+  ;; Taking Input (,)
+  ;; 
+
+  ;; Currently takes individual characters one at a time as prompted
+
+  (define (character-prompt)
+    "A prompt for obtaining one character as input."
+    (Lisp Char ()
+      (cl:format cl:*query-io* "Input a character: ")
+      (cl:force-output cl:*query-io*)
+      (cl:coerce (cl:read-line cl:*query-io*)
+                 'cl:character)))
+  
+  (define (take-input)
+    "Takes and stores a character at the input-index as an ascii code at
+the pointer."
+    (vec:set! (cell:read pointer)
+              (into (char:char-code (character-prompt)))
+              bf-array)
+    (current-bf-cell-value)))
+
+;;;
+;;; Parsing/Lexing
+;;;
+
+(coalton-toplevel
+
+  (define-type Cmd
+    BFRight
+    BFLeft
+    BFPlus
+    BFMinus
+    BFPrint
+    BFInput
+    (BFLoop (List Cmd)))
+
+  (declare split-string (String -> List String))
+  (define (split-string s)
```

Comment:
You might want to rewrite this to be tail recursive, otherwise your program will break on long inputs.